### PR TITLE
port scala3 macro 

### PR DESCRIPTION
**Pull request by timzaak**
ref #1068 
don't know how to implement `autoColumns`, it needs info out from macro method. I can not find any API to do this.

and  no test now.

### CODE REVIEW

**Comment by timzaak** [comment]
autoColumn passed my mock test case.
 tomorrow will try to pass CI

**Comment by timzaak** [comment]
wait https://github.com/lampepfl/dotty/issues/12076 issue closed

File: scalikejdbc-syntax-support-macro/src/main/scala3/scalikejdbc/SQLSyntaxSupportFactory.scala [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+          ${
+            Apply(Select.unique(New(TypeTree.of[A]), "<init>"), fields.map{case (name, typeTree) => {
+              val typeBinderTree = Implicits.search(TypeRepr.of[TypeBinder].appliedTo(typeTree.tpe)) match {
+                case result:ImplicitSearchSuccess => result.tree
+                case _ => report.throwError(s"could not find implicit of TypeBinder[${typeTree.show}]")
+              }
+              val exprs = typeTree.tpe.asType match {
+                case '[b] =>
+                  //generate must equal `implicitly[TypeBinder[FieldType]].apply(rs.underlying, scalikejdbc.autoColumns.camelToSnake(fieldName, nameConverters, useSnakeCaseColumnName)`
+                 '{${typeBinderTree.asExprOf[TypeBinder[b]]}.apply(rs.underlying, p(${Expr(name)}))}
```
Comment (timzaak) [inline scalikejdbc-syntax-support-macro/src/main/scala3/scalikejdbc/SQLSyntaxSupportFactory.scala:43]: If this is right?

**Comment by timzaak** [comment]
It has two bugs.

1. autoConstruct,SQLSyntaxSupportFactory don't support `class A(val a:xx, val b:xxx)`
2. SQLSyntaxSupportFactory don't support case class with no default value.

will try to fix later.

**Comment by timzaak** [comment]
don't know why CI break.

**Comment by timzaak** [comment]
@xuwei-k @seratch   I run test locally, and can not find the error of CI.

**Comment by seratch** [comment]
Hey @timzaak, thanks again for working on this. I'm afraid that build failures only with Scala 3 indicates the implementation is not fully compatible with the one for Scala 2. What do you think? 

I don't have the bandwidth to look into this now but am still wanting to merge your changes and release a new version for Scala 3!

**Comment by mblund** [comment]
Is the problem only due to some github workflow problems? Is the PR using the workflow definition from the main branch or something?

I cloned timzaaks repo and checked out the scala3-macro branch
```
> git clone https://github.com/timzaak/scalikejdbc.git
> cd scalikejdbc/
> git checkout scala3-macro
```
and then tried to run the same commands as the failing CI / test
```
> sbt -v SetScala3 root213/test root213/publishLocal
> ./test.sh
> ./scripted.sh
```
Then the failing tests above are working for me:
```
[info] autoNamedValues
[info] - should execute
[info] - should throw a compiler error if wrong column passed
[info] AutoSpec:
[info] autoConstruct
[info] - should execute
[info] - should throw a compiler error if unknown field name is decleared in excludes
```
Am I missing something obvious here? I am glad to try out something else if that helps you.

**Comment by xuwei-k** [comment]
https://github.com/scalikejdbc/scalikejdbc/blob/2906136fa625b106f5065003d11c43555498ac25/.github/workflows/ci.yml#L89

https://github.com/scalikejdbc/scalikejdbc/blob/2906136fa625b106f5065003d11c43555498ac25/.github/workflows/ci.yml#L57-L58

Scala 3 tests use MySQL config

**Comment by mblund** [comment]
oh, I see. I only had Postgres at hand, but after doing:
```
> cp -p scalikejdbc-core/src/test/resources/jdbc_postgresql.properties scalikejdbc-core/src/test/resources/jdbc.properties
```
before running the tests, I am getting the same errors as in the CI build
```
[info] AutoNamedValuesSpec:
[info] autoNamedValues
WARNING: An illegal reflective access operation has occurred
WARNING: Illegal reflective access by org.postgresql.jdbc.TimestampUtils (file:/Users/magnus/Library/Caches/Coursier/v1/https/repo1.maven.org/maven2/org/postgresql/postgresql/9.4.1212/postgresql-9.4.1212.jar) to field java.util.TimeZone.defaultTimeZone
WARNING: Please consider reporting this to the maintainers of org.postgresql.jdbc.TimestampUtils
WARNING: Use --illegal-access=warn to enable warnings of further illegal reflective access operations
WARNING: All illegal access operations will be denied in a future release
[info] - should execute *** FAILED ***
[info]   org.postgresql.util.PSQLException: The column name id was not found in this ResultSet.
[info]   at org.postgresql.jdbc.PgResultSet.findColumn(PgResultSet.java:2589)
[info]   at org.postgresql.jdbc.PgResultSet.getObject(PgResultSet.java:2580)
[info]   at org.apache.commons.dbcp2.DelegatingResultSet.getObject(DelegatingResultSet.java:763)
[info]   at org.apache.commons.dbcp2.DelegatingResultSet.getObject(DelegatingResultSet.java:763)
[info]   at scalikejdbc.DBConnectionAttributesWiredResultSet.getObject(DBConnectionAttributesWiredResultSet.scala:44)
[info]   at scalikejdbc.Binders$$anon$3.apply(Binders.scala:42)
[info]   at scalikejdbc.Binders$$anon$1.apply(Binders.scala:15)
[info]   at foo.AutoNamedValuesSpec$$anon$1.apply(AutoNamedValuesSpec.scala:41)
[info]   at foo.AutoNamedValuesSpec$$anon$1.apply(AutoNamedValuesSpec.scala:11)
[info]   at scalikejdbc.SQLSyntaxSupportImpl.apply(SQLSyntaxSupportImpl.scala:6)
...
[info] AutoSpec:
[info] autoConstruct
[info] - should execute *** FAILED ***
[info]   org.postgresql.util.PSQLException: The column name id was not found in this ResultSet.
[info]   at org.postgresql.jdbc.PgResultSet.findColumn(PgResultSet.java:2589)
[info]   at org.postgresql.jdbc.PgResultSet.getObject(PgResultSet.java:2580)
[info]   at org.apache.commons.dbcp2.DelegatingResultSet.getObject(DelegatingResultSet.java:763)
[info]   at org.apache.commons.dbcp2.DelegatingResultSet.getObject(DelegatingResultSet.java:763)
[info]   at scalikejdbc.DBConnectionAttributesWiredResultSet.getObject(DBConnectionAttributesWiredResultSet.scala:44)
[info]   at scalikejdbc.Binders$$anon$3.apply(Binders.scala:42)
[info]   at scalikejdbc.Binders$$anon$1.apply(Binders.scala:15)
[info]   at foo.AutoSpec$$anon$1.apply(AutoSpec.scala:38)
[info]   at foo.AutoSpec$$anon$1.apply(AutoSpec.scala:10)
[info]   at scalikejdbc.SQLSyntaxSupportImpl.apply(SQLSyntaxSupportImpl.scala:6)
```

**Comment by timzaak** [comment]
@mblund thanks, now I know how to debug it. `sbt -v SetScala3 "testOnly foo.AutoNamedValuesSpec"`, I will try to fix it later

File: scalikejdbc-syntax-support-macro/src/main/scala3/scalikejdbc/SQLSyntaxSupportFactory.scala [hunk 2]
```diff
@@ -37,7 +37,7 @@ object SQLSyntaxSupportFactory {
                 }
                 val exprs = typeTree.tpe.asType match {
                   case '[b] =>
-                   '{${typeBinderTree.asExprOf[TypeBinder[b]]}.apply(rs.underlying, p(${Expr(name)}))}
+                    '{${ypeBinderTree.asExprOf[TypeBinder[b]]}.apply(rs.underlying, rn.field(p(${Expr(name)})).value)}
```
Comment (mblund) [inline scalikejdbc-syntax-support-macro/src/main/scala3/scalikejdbc/SQLSyntaxSupportFactory.scala:40]: It looks like a typo here. ypeBinderTree without the 'T'

**Comment by seratch** [comment]
@timzaak @mblund @xuwei-k 
Thanks a lot for working on this PR & congrats on the completion of work here! 🎉 
It's time to release a new version for complete Scala 3 support 🚀

**Pull request merged**